src/audio/volume/volume.c

```diff
@@ -706,14 +737,24 @@ static int volume_prepare(struct processing_module *mod,
 
 	buffer_release(sink_c);
 
-	for (i = 0; i < cd->channels; i++) {
-		cd->volume[i] = cd->vol_min;
-		volume_set_chan(mod, i, cd->tvolume[i], false);
-		if (cd->volume[i] != cd->tvolume[i])
-			cd->ramp_finished = false;
-	}
-
-	volume_prepare_ramp(dev, cd);
+#if !CONFIG_COMP_VOLUME_PLAYBACK_START_FADE && !CONFIG_COMP_VOLUME_CAPTURE_START_FADE
+	volume_prepare_start_without_fade(mod);
+#elif CONFIG_COMP_VOLUME_PLAYBACK_START_FADE && CONFIG_COMP_VOLUME_CAPTURE_START_FADE
+	volume_prepare_start_with_fade(mod);
+#else
+#if CONFIG_COMP_VOLUME_PLAYBACK_START_FADE
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		volume_prepare_start_with_fade(mod);
+	else
+		volume_prepare_start_without_fade(mod);
+#endif /* CONFIG_COMP_VOLUME_PLAYBACK_START_FADE*/
+#if CONFIG_COMP_VOLUME_CAPTURE_START_FADE
+	if (dev->direction == SOF_IPC_STREAM_CAPTURE)
+		volume_prepare_start_with_fade(mod);
+	else
+		volume_prepare_start_without_fade(mod);
+#endif /* CONFIG_COMP_VOLUME_PLAYBACK_START_FADE*/
```

Comment:
those changes will make volume more complicated, we may need investigate and discuss first before take action.

Reply:
I tested all kconfig permutations so they work, but yes, there's more complexity.

Reply:
As this is not performance critical, I'd put the checks in an inline function and have just one call to volume_prepare_start_with_fade() and one to volume_prepare_start_without_fade().

```
static inline bool start_with_fade(direction)
{
#if CONFIG_COMP_VOLUME_PLAYBACK_START_FADE
	if (direction == SOF_IPC_STREAM_PLAYBACK)
                return true;
#endif 
#if CONFIG_COMP_VOLUME_CAPTURE_START_FADE
	if (direction == SOF_IPC_STREAM_CAPTURE)
                return true;
#endif 

        return false;
}
```

I think we can spare one conditional check in prepare().

Reply:
don't add new code if it is not must, our initial problem is high ramp mcps, now, with change to linear and optimize windows fade, we already get good results, as liam just mentioned in mail, we should keep module easy to read/understand, especially for volume such simply module.

If volume is complicated, how can we do for other real complicated modules.